### ISSUE TITLE
A possible divide by zero bug?

### ISSUE THREAD

**Issue by yiyuaner** [facebook/zstd#2591]
Hi,  in function `ZSTD_buildBlockEntropyStats_sequences`, we have the following [code](https://github.com/facebook/zstd/blob/f8afc6657393a5833ae9303e127cc76299893b20/lib/compress/zstd_compress.c#L3089):

```
size_t const nbSeq = seqStorePtr->sequences - seqStorePtr->sequencesStart;
...
stats = ZSTD_buildSequencesStatistics(seqStorePtr, nbSeq,...)
```
If `seqStorePtr->sequences == seqStorePtr->sequencesStart`, `nbSeq` will be equal to `0` and passed to the function `ZSTD_buildSequencesStatistics`.  Eventually, `nbSeq` will be used as the divisor of a division operation without checking.

Here is the full trace for triggering this problem (with links pointing to specific lines in the code):
[`ZSTD_buildBlockEntropyStats_sequences`](https://github.com/facebook/zstd/blob/f8afc6657393a5833ae9303e127cc76299893b20/lib/compress/zstd_compress.c#L3107)   --> [`ZSTD_buildSequencesStatistics`](https://github.com/facebook/zstd/blob/f8afc6657393a5833ae9303e127cc76299893b20/lib/compress/zstd_compress.c#L2412)  --> [`ZSTD_selectEncodingType`](https://github.com/facebook/zstd/blob/f8afc6657393a5833ae9303e127cc76299893b20/lib/compress/zstd_compress_sequences.c#L207) --> [`ZSTD_entropyCost`](https://github.com/facebook/zstd/blob/f8afc6657393a5833ae9303e127cc76299893b20/lib/compress/zstd_compress_sequences.c#L89), finally leading to a division in ZSTD_entropyCost:

``
        unsigned norm = (unsigned)((256 * count[s]) / total); // here total equals to nbSeq, which may equal to zero
``

**Comment by senhuang42** [comment]
Thanks for bringing this up. 

In practice, I don't believe that we can ever pass `nbSeq == 0` to `ZSTD_buildSequencesStatistics`. 
In the `ZSTD_entropyCompressSeqStore_internal()` entrypoint:
https://github.com/facebook/zstd/blob/f8afc6657393a5833ae9303e127cc76299893b20/lib/compress/zstd_compress.c#L2570-L2574

And the block splitter entry point:
https://github.com/facebook/zstd/blob/f8afc6657393a5833ae9303e127cc76299893b20/lib/compress/zstd_compress.c#L3498-L3500

Though this points out that it would be definitely be appropriate to add an `assert(nbSeq != 0)` .

**Comment by Cyan4973** [comment]
@senhuang42 's diagnosis is correct : a this point in the code, we have guarantees that a division by zero cannot happen, as such cases have already been filtered out before.
Such guarantees should nonetheless be materialized by an `assert()`.

@senhuang42 : on top of `nbSeq != 0` that you detected, we should also `assert( total != 0 )` directly in `ZSTD_entropyCost()`.

**Comment by yiyuaner** [comment]
Thanks for the response. I discovered another call sequence that seems suspicious:
https://github.com/facebook/zstd/blob/f8afc6657393a5833ae9303e127cc76299893b20/lib/compress/zstd_compress.c#L3780-L3791

In this function is it possible for the call `ZSTD_buildSeqStore(zc, src, srcSize)` to set  `zc->seqStore->sequences = zc->seqStore->sequencesStart` ? Looking into the code inside the function `ZSTD_buildSeqStore`, we first reset `seqStore`:
https://github.com/facebook/zstd/blob/f8afc6657393a5833ae9303e127cc76299893b20/lib/compress/zstd_compress.c#L2759

and it seems possible to leave the function `ZSTD_buildSeqStore` without touching `seqStore` again by entering this branch:
https://github.com/facebook/zstd/blob/f8afc6657393a5833ae9303e127cc76299893b20/lib/compress/zstd_compress.c#L2785-L2794

In function `ZSTD_compressBlock_targetCBlockSize`, if we call `ZSTD_compressBlock_targetCBlockSize_body` with  `zc->seqStore->sequences = zc->seqStore->sequencesStart`, it will in turn calls `ZSTD_compressSuperBlock`, `ZSTD_buildBlockEntropyStats`, and finally `ZSTD_buildBlockEntropyStats_sequences` (which leads to the divide by zero problem decribed above):

https://github.com/facebook/zstd/blob/f8afc6657393a5833ae9303e127cc76299893b20/lib/compress/zstd_compress.c#L3791
https://github.com/facebook/zstd/blob/f8afc6657393a5833ae9303e127cc76299893b20/lib/compress/zstd_compress.c#L3761
https://github.com/facebook/zstd/blob/f8afc6657393a5833ae9303e127cc76299893b20/lib/compress/zstd_compress_superblock.c#L555
https://github.com/facebook/zstd/blob/f8afc6657393a5833ae9303e127cc76299893b20/lib/compress/zstd_compress.c#L3142

Is this execution path feasible?

**Comment by terrelln** [comment]
I believe that that is fixed by #2592.
